keps/sig-storage/20190204-kubernetes-csi-release-process.md

```diff
@@ -0,0 +1,277 @@
+prefix (i.e. `v1.2.3`) is used for components that provide a stable
+API, like for example:
+- external-attacher
+- external-provisioner
+- external-snapshotter
+- cluster-driver-registrar
+- node-driver-registrar
+- csi-driver-host-path
+- csi-lib-utils
+
+Other components are internal utility packages that are
+not getting tagged:
+- csi-release-tools
+
+An additional suffix like `-rc1` can be used to denote a pre-release.
+
+The [hostpath example
+deployment](https://github.com/kubernetes-csi/csi-driver-host-path/tree/master/deploy)
+defines the exact release of each sidecar that is part of that
+deployment. When updating to newer sidecar releases, a new hostpath
+driver release is tagged with version numbers bumped according to the
+same rules as the individual components (major version bump when any
+of the sidecars had a major change, etc.).
+
+#### Release artifacts
+
+Tagging a component with a `v*` tag triggers a
+build for that component with that version. The output is primarily the container image
+for components that need to be deployed. Those images get published as
+`gcr.io/kubernetes-csi/<image>:<tag>`, which depends on [issue
+158](https://github.com/kubernetes/k8s.io/issues/158) getting resolved.
+
+Only binaries provided as part of a release with a semantic version tag without additional suffix
+should be considered
+production ready. Binaries are never going to be rebuilt, therefore an
+image like `csi-node-driver-registrar:v1.0.2` will always pull the
+same content and `imagePullPolicy: Always` can be omitted.
+
+Eventually auxiliary files like the RBAC rules that are included in
+each component might also get published in a single location, like
+`https://dl.k8s.io/kubernetes-csi/`, but this currently outside of
+this KEP.
+
+#### Release process
+
+1. A change is submitted against the master branch.
+
+1. A [Prow](https://github.com/kubernetes/test-infra/blob/master/prow/README.md)
+   job checks out the source of the modified component, rebuilds it and then
+   runs unit tests and E2E tests with it as defined below.
+
+1. Maintainers accept changes into the master branch.
+
+1. The same Prow job runs for master and repeats the checks. If it succeeds,
+   a new "canary" image is published for the component.
+
+1. In preparation for the release of a major new update, a feature freeze is
+   declared for the "master" branch and only changes relevant for that next
+   release are accepted.
+
+1. When all changes targeted for the release are in master, automatic
+   test results are okay and and potentially some more manual tests,
+   maintainers tag a new release directly on the master branch. This
+   can be a `-rc` test release or a normal release.
+
+1. Maintenance releases are prepared by creating a "release-X.Y" branch based on
+   release "vX.Y" and backporting relevant fixes from master. The same
+   prow job as for master also handles the maintenance branches.
```

Comment:
It's manual for every repository because (as far as I can tell) the Prow config cannot have wildcard entries for jobs that apply to multiple different repos. It is not necessary to have different jobs for each branch *if* we are strict about letting the repository control all relevant aspects of the actual execution inside the job (like Go version).

Basically we define an API between job and repository content (env variables provided by job and tools installed in the container image on the one hand, `.prow.sh` script provided by repository on the other) that we need to keep stable.

If we ever need to change this "API" in some incompatible way, we can still do it by setting up "legacy" jobs for older branches and changing to a newer job configuration for master.